File: src/main/java/no/nav/fo/veilarbregistrering/profilering/db/ProfileringRepositoryImpl.java
package no.nav.fo.veilarbregistrering.profilering.db;

import no.nav.fo.veilarbregistrering.profilering.Profilering;
import no.nav.fo.veilarbregistrering.profilering.ProfileringRepository;
import no.nav.sbl.sql.SqlUtils;
import no.nav.sbl.sql.where.WhereClause;
import org.springframework.jdbc.core.JdbcTemplate;

public class ProfileringRepositoryImpl implements ProfileringRepository {

    private final JdbcTemplate db;

    private final static String BRUKER_REGISTRERING_ID = "BRUKER_REGISTRERING_ID";

    private final static String BRUKER_PROFILERING = "BRUKER_PROFILERING";
    final static String PROFILERING_TYPE = "PROFILERING_TYPE";
    final static String VERDI = "VERDI";

    final static String ALDER = "ALDER";
    final static String ARB_6_AV_SISTE_12_MND = "ARB_6_AV_SISTE_12_MND";
    final static String RESULTAT_PROFILERING = "RESULTAT_PROFILERING";

    public ProfileringRepositoryImpl(JdbcTemplate db) {
        this.db = db;
    }

    @Override
    public void lagreProfilering(long brukerregistreringId, Profilering profilering) {
        SqlUtils.insert(db, BRUKER_PROFILERING)
                .value(BRUKER_REGISTRERING_ID, brukerregistreringId)
                .value(PROFILERING_TYPE, ALDER)
                .value(VERDI, profilering.getAlder())
                .execute();

        SqlUtils.insert(db, BRUKER_PROFILERING)
                .value(BRUKER_REGISTRERING_ID, brukerregistreringId)
                .value(PROFILERING_TYPE, ARB_6_AV_SISTE_12_MND)
                .value(VERDI, profilering.isJobbetSammenhengendeSeksAvTolvSisteManeder())
                .execute();

        SqlUtils.insert(db, BRUKER_PROFILERING)
                .value(BRUKER_REGISTRERING_ID, brukerregistreringId)
                .value(PROFILERING_TYPE, RESULTAT_PROFILERING)
                .value(VERDI, profilering.getInnsatsgruppe().getArenakode())
                .execute();
    }

    @Override
    public Profilering hentProfileringForId(long brukerregistreringId){
        return SqlUtils.select(db, BRUKER_PROFILERING, ProfileringMapper::map)
                .where(WhereClause.equals(BRUKER_REGISTRERING_ID, brukerregistreringId))
                .limit(3)
                .column("*")
                .execute();
    }
}

File: src/main/java/no/nav/fo/veilarbregistrering/db/profilering/ProfileringRepositoryImpl.kt
package no.nav.fo.veilarbregistrering.db.profilering

import no.nav.fo.veilarbregistrering.profilering.Innsatsgruppe
import no.nav.fo.veilarbregistrering.profilering.Profilering
import no.nav.fo.veilarbregistrering.profilering.ProfileringRepository
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import java.sql.ResultSet
import java.sql.SQLException

class ProfileringRepositoryImpl(private val db: NamedParameterJdbcTemplate) : ProfileringRepository {
    override fun lagreProfilering(brukerregistreringId: Long, profilering: Profilering) {
        val params1 = mapOf(
            "bruker_registrering_id" to brukerregistreringId,
            "type" to ALDER,
            "verdi" to profilering.alder
        )
        val params2 = mapOf(
            "bruker_registrering_id" to brukerregistreringId,
            "type" to ARB_6_AV_SISTE_12_MND,
            "verdi" to profilering.isJobbetSammenhengendeSeksAvTolvSisteManeder
        )
        val params3 = mapOf(
            "bruker_registrering_id" to brukerregistreringId,
            "type" to RESULTAT_PROFILERING,
            "verdi" to profilering.innsatsgruppe.arenakode
        )

        val sql = "INSERT INTO $BRUKER_PROFILERING ($BRUKER_REGISTRERING_ID, $PROFILERING_TYPE, $VERDI)" +
                " VALUES (:bruker_registrering_id, :type, :verdi)"

        db.update(sql, params1)
        db.update(sql, params2)
        db.update(sql, params3)
    }

    override fun hentProfileringForId(brukerregistreringId: Long): Profilering =
        db.query(
            "SELECT * FROM $BRUKER_PROFILERING WHERE $BRUKER_REGISTRERING_ID = :id" +
                " FETCH NEXT 3 ROWS ONLY",
            mapOf("id" to brukerregistreringId), mapper)!!


    companion object {
        private const val BRUKER_REGISTRERING_ID = "BRUKER_REGISTRERING_ID"
        private const val BRUKER_PROFILERING = "BRUKER_PROFILERING"
        const val PROFILERING_TYPE = "PROFILERING_TYPE"
        const val VERDI = "VERDI"
        const val ALDER = "ALDER"
        const val ARB_6_AV_SISTE_12_MND = "ARB_6_AV_SISTE_12_MND"
        const val RESULTAT_PROFILERING = "RESULTAT_PROFILERING"

        private val mapper: (ResultSet) -> Profilering = { rs ->
            try {
                Profilering().also {
                    while (rs.next()) {
                        when (rs.getString(PROFILERING_TYPE)) {
                            ALDER -> it.alder = rs.getInt(VERDI)

                            ARB_6_AV_SISTE_12_MND -> it.isJobbetSammenhengendeSeksAvTolvSisteManeder =
                                rs.getBoolean(VERDI)

                            RESULTAT_PROFILERING -> it.innsatsgruppe = Innsatsgruppe.of(rs.getString(VERDI))
                        }
                    }
                }
            } catch (e: SQLException) {
                throw RuntimeException(e)
            }
        }
    }
}